src/shims/native_lib/mod.rs

```diff
@@ -31,14 +39,12 @@ trait EvalContextExtPriv<'tcx>: crate::MiriInterpCxExt<'tcx> {
         dest: &MPlaceTy<'tcx>,
         ptr: CodePtr,
         libffi_args: Vec<libffi::high::Arg<'a>>,
-    ) -> CallResult<'tcx> {
+    ) -> trace::CallResult<'tcx> {
         let this = self.eval_context_mut();
-        //#[cfg(target_os = "linux")]
-        //let alloc = this.machine.allocator.as_ref().unwrap();
+        let alloc = this.machine.allocator.as_ref().unwrap();
 
         // SAFETY: We don't touch the machine memory past this point.
-        //#[cfg(target_os = "linux")]
-        //let (guard, stack_ptr) = unsafe { Supervisor::start_ffi(alloc) };
+        let (guard, stack_ptr) = unsafe { Supervisor::start_ffi(alloc) };
```

Comment:
Just to make sure this doesn't get lost, this should become just `let guard = ` with a new `FfiCallGuard` type or so that stores the `stack_ptr` internally.

Or, would it be better to make this a callback-based API, and thus fully exclude any chance of the user forgetting to call `end_ffi`?
```rust
unsafe fn do_ffi(
  alloc: &Rc<RefCell<IsolatedAlloc>>,
  f: impl FnOnce(),
);
```
The only annoying thing is this puts the entire closure inside the unsafe block, but the closure is already half-unsafe so that's not such a big deal.

Reply:
Ah, we also need the closure to return something:
```rust
unsafe fn do_ffi<R>(
  alloc: &Rc<RefCell<IsolatedAlloc>>,
  f: impl FnOnce() -> InterpResult<'tcx, R>,
) -> InterpResult<'tcx, (R, Option<MemEvents>)>
```
That then also avoids the slightly awkward `break 'ret` stuff we currently use, we can just `return` from the closure.

i like this, let's go for this. :)